engine/src/cmd/collide_map.cpp

```diff
@@ -36,27 +39,23 @@ void CollideArray::erase(iterator target) {
     count -= 1;
     if (target >= this->begin() && target < this->end()) {
         target->radius = 0;
-        target->ref.unit = NULL;
+        target->ref.unit = nullptr;
         size_t diff = (target - this->begin());
         if (this->unsorted.size() > diff) {
             //for secondary collide arrays that have no unsorted array
             iterator tmp = &*(this->unsorted.begin() + diff);
             tmp->radius = 0;
-            tmp->ref.unit = NULL;
+            tmp->ref.unit = nullptr;
         }
         return;
-    } else if (target == NULL) {
+    } else if (target == nullptr) {
         return;
     } else {
-        CollidableBackref *targ = static_cast< CollidableBackref * > (&*target);
-        std::list<CollidableBackref> *targlist = &toflattenhints[targ->toflattenhints_offset];
-        std::list<CollidableBackref>::iterator endlist = targlist->end();
-        for (std::list<CollidableBackref>::iterator i = targlist->begin(); i != endlist; ++i) {
-            if (&*i == target) {
-                targlist->erase(i);
-                return;
-            }
-        }
+        CollidableBackref *targ = vega_dynamic_cast_ptr<CollidableBackref>(&*target);
```

Comment:
I don't understand this code at all. Can we maybe split this into a separate PR?

Reply:
We were doing a `static_cast` to pointer-to-CollidableBackref, without any type checking, where we should have been doing a `dynamic_cast` in order to verify that the actual type (at runtime) is correct.

Does that explain it sufficiently? Or do you want to go ahead and split this part out into its own PR?

Reply:
That's fine.